wordlists: init at unstable-2020-11-23

Pamplemousse opened the pull request:
Signed-off-by: Pamplemousse <xav.maso@gmail.com>

###### Motivation for this change

Wordlists are a must-have for pentesting.
I thought it would be nice to package some, for ease of use.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Pamplemousse committed:
security-wordlists: init at unstable-2020-11-23

Signed-off-by: Pamplemousse <xav.maso@gmail.com>

Pamplemousse commented:
I am quite unsure if this fits `nixpkgs`, or if I should put that elsewhere (NUR).
Thoughts and feedback appreciated :)

sikmir commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+  version = "1.0-RC1";
+
+  src = fetchurl {
+    url = "https://sourceforge.net/projects/dirbuster/files/DirBuster%20%28jar%20%2B%20lists%29/${version}/${pname}-${version}.zip";
```
```suggestion
    url = "mirror://sourceforge/projects/dirbuster/files/DirBuster%20%28jar%20%2B%20lists%29/${version}/${pname}-${version}.zip";
```

Pamplemousse replied:
This does not work: it fetches an HTML document, and not the ZIP file I am expecting.
How can I browse the content of `sourceforge`'s mirror to locate the resource I am looking for?

sikmir replied:
I've updated url, try it now.

sikmir replied:
`mirror://sourceforge/` is:
https://github.com/NixOS/nixpkgs/blob/dc0d766f2dc06a01e2e3b35d72122cc22b631a35/pkgs/build-support/fetchurl/mirrors.nix#L19-L27

iblech commented:
Oh, nice! I wanted to have a couple wordlists available in the past, but was too lazy to package them. We also ship john the ripper, it might even make sense to mark one of these wordlists as dependencies of john, hence I do believe that these wordlists are useful additions to nixpkgs. Thank you for your work, very much appreciated.

jonringer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28894,5 +28894,9 @@ in
 
   lc3tools = callPackage ../development/tools/lc3tools {};
 
+  wordlists = callPackage ../tools/security/wordlists/default.nix {
+    lists = [ "dirbuster" "nmap" "rockyou" "seclists" "wfuzz" ];
+  };
```
I think this should be a package scope, this diff shows a pretty minimal example of how to do so:
https://github.com/NixOS/nixpkgs/commit/632c4f2c9ba1f88cd5662da7bedf2ca5f0cda4a9

jonringer replied:
and then make it so that wordlists exposes a `withLists` helper similar to python's `withPackages` function:
```
  wordlists.withLists(l: with l; [ dirbuster nmap rockyou seclists wfuzz ])
```

also avoids things being "stringly" typed

jonringer replied:
the `withLists` can be implemented in a later PR, just demonstrating why a package scope would be beneficial

Pamplemousse replied:
Hmmm, I am not too sure how that would work...

Especially with the script that is currently created (to create the `wordlists` and `wordlists_path` commands).

From the example you shared, and reading more in the current `nixpkgs`, I have the impression that `cinnamon` is "just" a bundle of packages.
Right now, I was seeing `wordlists` as a 1) Bundle of wordlists, 2) Handy commands to access them.

Pamplemousse replied:
@jonringer, I am really not familiar with this pattern...

I kinda refactored the original code (first commit); And tried to follow your hint, although I kept it separate (third commit).
Is that the kind of things you were hoping to see?

Now I am having a look on how to implement some `withLists`...
And probably have the `extraLists` feature to follow a similar idea.

jonringer commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+, unzip
+}:
+
+stdenv.mkDerivation  rec {
+  pname = "DirBuster";
+  version = "1.0-RC1";
+
+  src = fetchurl {
+    url = "mirror://sourceforge/dirbuster/DirBuster%20%28jar%20%2B%20lists%29/${version}/${pname}-${version}.zip";
+    sha256 = "sha256-2oDRe9NjvGDT5yFqPEMylhfL1iCsVeVeJ1G9MXfQnqE=";
+  };
+
+  unpackCmd = "${unzip}/bin/unzip $curSrc";
+
+  dontConfigure = true;
+
+  installPhase = ''
+    mkdir -p $out/share/dirbuster
+    install -m 444 directory-list-*.txt $out/share/dirbuster/
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = "https://sourceforge.net/projects/dirbuster/";
+    license = licenses.gpl2;
+    maintainers = with maintainers; [ pamplemousse ];
+  };
+}
```
at lot of these could probably converted over to a `fetchzip`, and just use `postFetch` to unpack the correct contents

Pamplemousse replied:
Do you mean `postUnpack`?

I just tried using `postUnpack`, and it seems that `$out` is not yet present...
What about I just replace the `fetchurl` by a `fetchzip`, and get rid of the `unpackCmd`?

jonringer replied:
no, `postFetch`

it's special to fetchurl, which is used by fetchzip and fetchFromGit{Hub,Lab}

jonringer commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,22 @@
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = wfuzz.meta.homepage;
+    license = wfuzz.meta.license;
```
```suggestion
    inherit (wfuzz.meta) license homepage;
```

jonringer commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,28 @@
+  pname = "rockyou";
+  version = "0.3-1kali3";
+
+  src = fetchFromGitLab {
+    group = "kalilinux";
+    owner = "packages";
+    repo = "wordlists";
+    rev = "debian/${version}";
+    sha256 = "sha256-viOd8iKLTiHrPF0azICp/F16JrPUfo4kbcGUPD2/cRs=";
+  } + "/rockyou.txt.gz";
+
+  unpackCmd = "mkdir src && gzip -dc $curSrc > src/rockyou.txt";
```
```suggestion
  src = fetchFromGitLab {
    group = "kalilinux";
    owner = "packages";
    repo = "wordlists";
    rev = "debian/${version}";
    sha256 = "sha256-viOd8iKLTiHrPF0azICp/F16JrPUfo4kbcGUPD2/cRs=";
  };

  unpackCmd = "mkdir src && gzip -dc ${src}/rockyou.txt.gz"> src/rockyou.txt";
```

jonringer commented on pkgs/tools/security/wordlists/seclists.nix:
```diff
@@ -0,0 +1,28 @@
+    mkdir -p $out/share/SecLists
+    cp -R Discovery Fuzzing IOCs Miscellaneous Passwords Pattern-Matching Payloads Usernames Web-Shells \
+      $out/share/SecLists
+    find $out/share/SecLists -name "*.md" -exec rm '{}' \;
```
```suggestion
    find $out/share/SecLists -name "*.md" -delete
```

jonringer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28894,5 +28894,9 @@ in
 
   lc3tools = callPackage ../development/tools/lc3tools {};
 
+  wordlists = callPackage ../tools/security/wordlists/default.nix {
```
also, you may want to change the name, it's already in use, which is why we have the 5k+ rebuilds

```
pkgs/development/libraries/cracklib/default.nix
3:, wordlists ? [ (fetchurl {
26:    ln -vs ${toString wordlists} dicts/
38:    license = licenses.lgpl21; # Different license for the wordlist: http://www.openwall.com/wordlists
```

jonringer replied:
this is later used by pam, which why it's almost a full stdenv rebuild

```
pkgs/os-specific/linux/pam/default.nix
1:{ stdenv, buildPackages, fetchurl, fetchpatch, flex, cracklib, db4 }:
31:  buildInputs = [ cracklib db4 ];
```

duraki commented:
Wouldn't be better to include [danielmiessler/SecLists](github.com/danielmiessler/SecLists) gitdir? (ie. w/ `--depth=1`). Much larger sample.

Pamplemousse commented:
@duraki, It is already included in the PR!
  - [the expression for SecLists](https://github.com/NixOS/nixpkgs/pull/104712/files#diff-22505fbd5835058407d3ba6947c91c75c28a74ed24dcf1869b6f037b747a7dd8)
  - [the inclusion in `security-wordlists`](https://github.com/NixOS/nixpkgs/pull/104712/files#diff-b158abb8d6e9bbe00a538089888d7836c1cb7522fd98286032ff5edc834890c7R12)

duraki commented:
Thanks a lot, missed that. Hopefully this PR will get accepted soon. +paging @jonringer

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/435

jonringer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28936,5 +28936,7 @@ in
 
   lc3tools = callPackage ../development/tools/lc3tools {};
 
+  security-wordlists = callPackage ../tools/security/wordlists/default.nix {};
```
normal convention to drop the default.nix explicitly
```suggestion
  security-wordlists = callPackage ../tools/security/wordlists {};
```

jonringer replied:
also, I'm not a big fan of new attrs at the end of `all-packages.nix`, usually leads to a lot of merge conflicts.

jonringer reviewed (commented):
mostly looks good to me, may just need some documentation, so people know how to use it.

Maybe a  `doc/misc/pen-testing/wordlists.md` which shows some usage:
```
(security-wordlists.withLists (lists: with lists; [ nmap wfuzz seclists ]))
```

Pamplemousse commented:
What about putting it here: https://nixos.org/manual/nixpkgs/stable/#chap-packages, under a `Security Wordlists` title?

I am a bit reluctant creating a `misc` section, containing a single subsection `pentesting` for a single package...

Pamplemousse committed:
security-wordlists: Add documentation

Signed-off-by: Pamplemousse <xav.maso@gmail.com>

jonringer commented:
cc @SuperSandro2000 , looks okay to me

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29197,6 +29197,8 @@ in
 
   prow = callPackage ../applications/networking/cluster/prow { };
 
+  security-wordlists = callPackage ../tools/security/wordlists {};
```
```suggestion
  security-wordlists = callPackage ../tools/security/wordlists { };
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+{ fetchFromGitHub
+, pkgs
```
```suggestion
, wfuzz
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+
+in stdenv.mkDerivation rec {
+  pname = "wfuzz";
+  version = pkgs.wfuzz.version;
```
```suggestion
  version = wfuzz.version;
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+  pname = "wfuzz";
+  version = pkgs.wfuzz.version;
+
+  src = pkgs.wfuzz.src + "/wordlist";
```
```suggestion
  src = wfuzz.src + "/wordlist";
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+  passthru = { inherit shared; };
+
+  meta = with stdenv.lib; {
+    inherit (pkgs.wfuzz.meta) license homepage;
```
```suggestion
    inherit (wfuzz.meta) license homepage;
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+
+  passthru = { inherit shared; };
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+{ fetchFromGitHub
```
```suggestion
{ lib
, fetchFromGitHub
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/seclists.nix:
```diff
@@ -0,0 +1,33 @@
+
+  passthru = { inherit shared; };
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/seclists.nix:
```diff
@@ -0,0 +1,33 @@
+{ fetchFromGitHub
```
```suggestion
{ lib
, fetchFromGitHub
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,33 @@
+
+  passthru = { inherit shared; };
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "sha256-viOd8iKLTiHrPF0azICp/F16JrPUfo4kbcGUPD2/cRs=";
+  };
+
+  unpackCmd = "mkdir src && gzip -dc ${src}/rockyou.txt.gz > src/rockyou.txt";
```
```suggestion
  unpackCmd = ''
    mkdir src
    gzip -dc ${src}/rockyou.txt.gz > src/rockyou.txt
  '';
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,33 @@
+{ fetchFromGitLab
```
```suggestion
{ lib
, fetchFromGitLab
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/nmap.nix:
```diff
@@ -0,0 +1,33 @@
+
+  passthru = { inherit shared; };
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/nmap.nix:
```diff
@@ -0,0 +1,33 @@
+  passthru = { inherit shared; };
+
+  meta = with stdenv.lib; {
+    inherit (pkgs.nmap.meta) homepage license;
```
```suggestion
    inherit (nmap.meta) homepage license;
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/nmap.nix:
```diff
@@ -0,0 +1,33 @@
+{ fetchFromGitHub
+, pkgs
```
```suggestion
, nmap
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/nmap.nix:
```diff
@@ -0,0 +1,33 @@
+{ fetchFromGitHub
```
```suggestion
{ lib
, fetchFromGitHub
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+
+  meta = with lib; {
+    homepage = "https://sourceforge.net/projects/dirbuster/";
+    license = licenses.gpl2;
```
gpl2Only or gpl2Plus?

Pamplemousse replied:
I am not too sure how to interpret that at https://sourceforge.net/projects/dirbuster/ ...
I would say `gpl2Only`?

buckley310 replied:
I noticed the software is GPL, but the wordlists included with dirbuster themselves seem to be a different license. They all have copyright headers with "Creative Commons Attribution-Share Alike 3.0"

SuperSandro2000 replied:
The software itself it lgpl2only. If the lists have that header please mark them as that.

Pamplemousse replied:
Oh, good catch @buckley310 !

SuperSandro2000 reviewed (changes requested):
@ofborg eval

Pamplemousse commented:
@SuperSandro2000 , I am having a hard time not using `pkgs`:
For example, if I replace the occurrences of `pkgs.wfuzz` in `wfuzz.nix`, the I get the following:
```bash
nix-shell -p 'security-wordlists.withLists(l: with l; [ wfuzz ])'
error: --- EvalError ----------------------------------------------------------------------------------------------- nix-shell
at: (11:3) in file: /home/pamplemousse/Workspace/tools/nixpkgs/pkgs/tools/security/wordlists/wfuzz.nix

    10|   pname = "wfuzz";
    11|   version = wfuzz.version;
      |   ^
    12| 

infinite recursion encountered
```

SuperSandro2000 commented:
> @SuperSandro2000 , I am having a hard time not using `pkgs`:
> For example, if I replace the occurrences of `pkgs.wfuzz` in `wfuzz.nix`, the I get the following:
> 
> ```shell
> nix-shell -p 'security-wordlists.withLists(l: with l; [ wfuzz ])'
> error: --- EvalError ----------------------------------------------------------------------------------------------- nix-shell
> at: (11:3) in file: /home/pamplemousse/Workspace/tools/nixpkgs/pkgs/tools/security/wordlists/wfuzz.nix
> 
>     10|   pname = "wfuzz";
>     11|   version = wfuzz.version;
>       |   ^
>     12| 
> 
> infinite recursion encountered
> ```

It is referencing itself. If you can't get it to use the version number from wfuzz without pigs than we can leave it like this.

Pamplemousse committed:
security-wordlists: Refactoring

Don't pass `pkgs` to expressions, but the relevant specific package.

Signed-off-by: Pamplemousse <xav.maso@gmail.com>

Pamplemousse commented:
@SuperSandro2000, AFAIR, this is why I did it originally.
However, I had an idea to make this happen; See latest commit.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 104712` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>security-wordlists</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
security-wordlists:
<details>


Package is missing a license

Near pkgs/tools/security/wordlists/default.nix:71:7:

```
   |
71 |       meta = with lib; {
   |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/license-missing.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/security/wordlists/default.nix:43:7:

```
   |
43 |       installPhase = ''
   |       ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
</details>

SuperSandro2000 commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+}:
+
+let
+  pname = "DirBuster";
+  version = "1.0-RC1";
+
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+  shared = "dirbuster";
+
+in fetchzip rec {
+  name = "${pname}-${version}";
```
```suggestion
  name = "DirBuster-1.0-RC1";
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+in fetchzip rec {
+  name = "${pname}-${version}";
+
+  url = "mirror://sourceforge/dirbuster/DirBuster%20%28jar%20%2B%20lists%29/${version}/${pname}-${version}.zip";
```
```suggestion
  url = "mirror://sourceforge/dirbuster/DirBuster%20%28jar%20%2B%20lists%29/${version}/${name}.zip";
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,29 @@
+
+  postFetch = ''
+    mkdir -p $out/share/dirbuster
+    unzip -j $downloadedFile ${pname}-${version}/directory-list-\*.txt -d $out/share/${shared}
```
```suggestion
    unzip -j $downloadedFile ${name}/directory-list-\*.txt -d $out/share/${shared}
```

SuperSandro2000 commented on pkgs/tools/security/wordlists/nmap.nix:
```diff
@@ -0,0 +1,34 @@
+
+in stdenv.mkDerivation {
+  pname = "nmap.lst";
+  version = "unstable-2020-10-19";
```
Is this still under the free license or the unfree?

Pamplemousse replied:
Oh wow, I am just learning about the `nmap` license update...

As for the list of passwords, it does not seem to have been updated (prior to the license change) since 2012! So IMHO, we can go with the `free` version.

I updated the commit (and related sha), to specifically pin a commit prior to the license change.

SuperSandro2000 reviewed (commented):
I think the packages are good to go but I am not sure about the nix functions.

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,78 @@
+    ];
+
+  scopedWordlists = lib.makeScope pkgs.newScope (self: with self;
+    let _packageWithLinkCommandFrom = acc: datum:
+      # Some wordlists are based on an already existing `nixpkgs` package; Pass it as parameter of the expression.
+      let parameters = {}
+        // lib.optionalAttrs ((builtins.elem "needsPackage" (builtins.attrNames datum)) && datum.needsPackage) { ${datum.name} = pkgs.${datum.name}; };
```
```suggestion
    let _packageWithLinkCommandFrom = acc: datum@{ needsPackage ? false, ... }:
      # Some wordlists are based on an already existing `nixpkgs` package; Pass it as parameter of the expression.
      # I'm not sure if the `inherit` usage below works as expected, but should be tried.
      let parameters = lib.optionalAttrs needsPackage { inherit (pkgs) "${datum.name}"; };
```

Pamplemousse replied:
Sadly this is illegal!

```nix
error: --- ParseError ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ nix-shell
at: (21:72) in file: /home/pamplemousse/Workspace/tools/nixpkgs/pkgs/tools/security/wordlists/default.nix

    20|       # I'm not sure if the `inherit` usage below works as expected, but should be tried.
    21|       let parameters = lib.optionalAttrs needsPackage { inherit (pkgs) "${datum.name}"; };
      |                                                                        ^

dynamic attributes not allowed in inherit
```

zseri replied:
ok, the rest of the suggestion still applies:
* `{} //` is unnecessary
* the whole builtin mashup can be replaced by extracting out of the argument

zseri replied:
```suggestion
    let _packageWithLinkCommandFrom = acc: datum@{ needsPackage ? false, ... }:
      # Some wordlists are based on an already existing `nixpkgs` package; Pass it as parameter of the expression.
      let parameters = lib.optionalAttrs needsPackage { ${datum.name} = pkgs.${datum.name}; };
```

zseri replied:
remains to be fixed.

zseri replied:
still not fixed.

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,78 @@
+        maintainers = with maintainers; [ pamplemousse ];
+      };
+    };
+in packageConstructor rec {
```
```suggestion
in packageConstructor {
```

zseri replied:
huh, the inner expression does not access `withLists`, so `rec` is unnecessary.

zseri commented on pkgs/tools/security/wordlists/dirbuster.nix:
```diff
@@ -0,0 +1,27 @@
+{ fetchzip
+, lib
+}:
```
```suggestion
{ fetchzip, lib }:
```

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,78 @@
+in let
+  # Given an array of wordlist packages,
+  # return an array of bash commands to create symlinks to the files shared by each package.
+  allLinksCommandFor = wordlistPackages:
```
I think you could use [`symlinkJoin`](https://github.com/NixOS/nixpkgs/blob/6af93b840564dd307f92d237569bad947ebf25e2/pkgs/build-support/trivial-builders.nix#L282) here instead ([usage description](https://nixos.wiki/wiki/Nix_Cookbook#Wrapping_packages))

zseri commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+  shared = "wfuzz";
+
+in stdenv.mkDerivation rec {
+  pname = "wfuzz";
```
I don't think this `pname` is correct... Shouldn't it be `wfuzz.lst`?

Pamplemousse replied:
I don't see why: everything is scoped under `wordlists`.
For consistency with the rest, I changed `nmap`'s `pname`.

zseri replied:
ah ok.

zseri replied:
It might be appropriate if this would be just a separate output from `wfuzz`, but this depends on the size of the wordlist (I'm not sure what's an appropriate cutoff point; the list seems to have a size of around ~1,2MiB).

zseri commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+let
+  shared = "wfuzz";
+
+in stdenv.mkDerivation rec {
```
```suggestion
in stdenv.mkDerivation {
```

zseri commented on pkgs/tools/security/wordlists/nmap.nix:
```diff
@@ -0,0 +1,34 @@
+  dontUnpack = true;
+
+  installPhase = ''
+    mkdir -p $out/share
```
This mkdir is unnecessary, because `install` is called with `-D` below.
```suggestion
```

zseri commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,37 @@
+  '';
+
+  installPhase = ''
+    mkdir -p $out/share
```
```suggestion
```

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,78 @@
+{ lib
```
I think this dependency list should be sorted (only `pkgs` needs to be moved).

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,78 @@
+    lib.concatStringsSep "\n" (builtins.map (x: "ln -s ${x}/share/${x.passthru.shared} $out/share/${x.passthru.shared}") wordlistPackages);
+
+in let
+  packageConstructor = { withLists ? null, wordlistPackages ? [] }:
```
This handling of `withLists` irritates me a bit. Why is it included passively (via passthru) in the derivation?

Pamplemousse replied:
> Why is it included passively (via passthru) in the derivation?

Because I didn't (and still don't) know how to have `security-wordlists` to both have `.pkgs.<...>` and `.withLists`.

There are two usages for this package:
  * one as a package to use, via `wordlists` and `wordlists_path` shell commands
  * one to allow a specific wordlist to be used a requirement in a derivation: e.g. `buildInputs = [ security-wordlists.pkgs.nmap ]`

> Why not simply returning a function, which takes withLists as input.

I am not too certain I understand what you propose, but it seems that having a function to take `withLists` as input would break the second case.

zseri commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,37 @@
+    repo = "wordlists";
+    rev = "debian/${version}";
+    sha256 = "sha256-viOd8iKLTiHrPF0azICp/F16JrPUfo4kbcGUPD2/cRs=";
+  };
```
For congruence with the other descriptions, it might be a good idea to put `+ "rockyou.txt.gz"` after `}`:
```suggestion
  } + "rockyou.txt.gz";
```
Then, do the opposite of this change to the `gzip` unpacking line.

zseri commented:
hm... it's a bit irritating that you mark stuff as resolved before pushing a commit which solves it.

Pamplemousse commented:
@zseri Oh sorry, it helps me keep track of what I have left to look at...
I will re-request your review once I push :)

zseri commented:
ok, thanks.

zseri commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,26 @@
+{ lib
+, stdenv
+, wfuzz
+}:
```
```suggestion
{ lib, stdenv, wfuzz }:
```

Pamplemousse committed:
security-wordlists: Refactoring

Use `symlinkJoin` instead of managing the links manually.

Signed-off-by: Pamplemousse <xav.maso@gmail.com>

Pamplemousse commented:
@zseri
I haven't addressed your comments on the style of the parameters of the function.
Unless there is a style guide that has explicit rules about that, it's not better than any other opinion...

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,74 @@
+        maintainers = with maintainers; [ pamplemousse ];
+      };
+    };
+in packageConstructor {
+  withLists = f: packageConstructor { wordlistPackages = f scopedWordlists; };
+}
```
```suggestion
in {
  pkgs = scopedWordlists;
  withLists = f: packageConstructor { wordlistPackages = f scopedWordlists; };
}
```
I think this might work, and gets rid of the otherwise unused derivation.

zseri commented:
I would want the code at https://github.com/NixOS/nixpkgs/pull/104712/files#r589680317 to be adjusted. That's basically my biggest concern. If there are evaluation errors, which render my suggestion unusable, please tell me.

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,74 @@
+        chmod +x $out/bin/wordlists_path
+      '';
+
+      passthru = {
+        inherit withLists;
+        pkgs = scopedWordlists;
+      };
```
After/when https://github.com/NixOS/nixpkgs/pull/104712/files#r589834021 is applied, then this bunch of `passthru` can be removed.

zseri commented on pkgs/tools/security/wordlists/wfuzz.nix:
```diff
@@ -0,0 +1,21 @@
+
+stdenv.mkDerivation {
+  pname = "wfuzz";
+  version = wfuzz.version;
```
```suggestion
  inherit (wfuzz) version;
```

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,74 @@
+
+      propagatedBuildInputs = [ tree ];
+
+      symlinks = symlinkJoin {
```
NOTE: you could replace the `stdenv.mkDerivation` above with `symlinkJoin`, as symlinkJoin should also accept the classic derivation parameters (this package has nothing to really compile, so it won't need the compile stuff included in `stdenv`).

Pamplemousse replied:
I tried to, but the `installPhase`, during which the `wordlists` and `wordlists_path` commands are defined, does not seem to be running...

zseri replied:
You would need to replace `installPhase` with `postBuild`, like I've done in https://github.com/Pamplemousse/nixpkgs/pull/1

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,74 @@
+
+in let
+  packageConstructor = { withLists ? null, wordlistPackages ? [] }:
+    stdenv.mkDerivation rec {
```
```suggestion
    symlinkJoin rec {
```

zseri commented:
If it's easier for you, I could also open a PR on you PR which applies most of my suggestions; I'll try to split it in appropriate parts, with the important stuff at front, and stuff with more divided opinions (e.g. style stuff) at the end, so that I can be cut off at a point which is reasonable for you.

fogti committed:
security-wordlists / wfuzz: move wordlists output into wfuzz

zseri commented:
I'm unsure if `SecLists` really fits as a "wordlist", as it contains various payloads and stuff, which aren't "wordlists" in a strict sense. It contains a folder `Passwords` which would match the sense "wordlist", but the rest, in large parts ( I haven't looked into everything) doesn't. SecLists should be probably split into different outputs, or the scope reduced (e.g. only some subdirectories exported).

Pamplemousse commented:
Yeah, SecLists is a big blob of things.

Although, I don't quite agree with you, there are plenty of other things that would fit what is expected from a "wordlist" in there: `Usernames`, `Discovery/` (similar to what `dirbuster` has), `Miscellaneous/` (security questions answers, top domains) to name a few.

"wordlist" is not just for password cracking, it's quite useful to enumerate things in general, or to feed to a grammar based buzzer...

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -40,25 +36,25 @@ in let
+        paths = wordlistPackages;
+      };
 
+      installPhase = ''
```
```suggestion
      postBuild = ''
```

nixinator commented:
Squash those commits by rebasing into one?

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+      dontUnpack = true;
+
+      propagatedBuildInputs = [ tree ];
```
why is this necessary?

Pamplemousse replied:
Because `tree` is a "runtime" dependency of this package; See 7 lines below.

zseri replied:
My reasoning for this:
 * `tree` is referenced directly in the builder, so we don't need to specify it at build-time.
 * `tree` is present in cleartext in `$out`, so nix will find that reference and correctly deal with it.
Thus we don't need to specify it in `propagatedBuildInputs`

zseri commented on pkgs/tools/security/wordlists/rockyou.nix:
```diff
@@ -0,0 +1,31 @@
+    repo = "wordlists";
+    rev = "debian/${version}";
+    sha256 = "sha256-viOd8iKLTiHrPF0azICp/F16JrPUfo4kbcGUPD2/cRs=";
+  } + "/rockyou.txt.gz";
+
+  unpackCmd = ''
+    mkdir src
+    gzip -dc ${src} > src/rockyou.txt
```
we do a string interpolation anyway, so it might be a good idea to do this:
```suggestion
  };

  unpackCmd = ''
    mkdir src
    gzip -dc ${src}/rockyou.txt.gz > src/rockyou.txt
```

Pamplemousse replied:
I prefer the original one: it makes it explicit that the source is only the `.txt.gz` file.

Pamplemousse commented:
@nixinator This can be done at "merge" time through GH's UI...

Pamplemousse commented:
@SuperSandro2000 @jonringer any chance this can be merged?

This PR has been reviewed by many people already;
I have fixed things as soon as possible every time;
But the PR gets forgotten every other month;
And now, it seems to mostly attract nitpicking when I ask for "review"...

jonringer commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,68 @@
+      };
+    };
+in {
+  pkgs = scopedWordlists;
```
Sorry for pro-longing this PR, but once it gets merged, then I would like avoid breaking people who use it if we decide to pivot.

I think this would create paths like `security-wordlists.pkgs.wfuzz`. Wouldn't it we want a scope more like `security-wordlists.wfuzz`... I guess if we have more abstractions at the `security-wordlists` package set level then maybe it would warrant another attr (e.g. `security-wordlists.all`, etc).

Other thing, is that maybe we should call it `lists` instead of `pkgs`? `pkgs` sound more like binaries/libraries. At least in my head.

I don't feel super strong about this, just want to make sure it's what we would like moving forward and "written into stone"

Pamplemousse replied:
> I think this would create paths like `security-wordlists.pkgs.wfuzz`

Yes, it does.

> Wouldn't it we want a scope more like `security-wordlists.wfuzz`...

Not sure that would fit the use cases highlighted in https://github.com/Pamplemousse/nixpkgs/blob/wordlists/doc/builders/packages/security-wordlists.section.md#usage ...

And AFAIR, I wanted it to have some consistency with other parts of `nixpkgs`;
In particular, being familiar `python`, I was aiming to "replicate" its structure:
  * `python` --> `security-wordlists`, a package. The latter builds two `bin/`: `wordlists` and `wordlists_path`;
  * `python.pkgs` --> `security-wordlists.pkgs`, a set of "sub-packages". The latter being the lists per se.

> Other thing, is that maybe we should call it lists instead of pkgs? pkgs sound more like binaries/libraries.

Why not. But again, what about consistency across `nixpkgs` (https://github.com/FRidh/rfcs/blob/83/rfcs/0083-common-interface-package-sets.md)?
And in the end, a wordlist is a package (see for example to import in an expression: https://github.com/Pamplemousse/nixpkgs/blob/wordlists/doc/builders/packages/security-wordlists.section.md#in-an-expression)...

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,68 @@
+}:
+
+let
+  _wordlistData =
+    [ { name = "dirbuster"; expressionPath = ./dirbuster.nix; }
+      { name = "nmap"; expressionPath = ./nmap.nix; needsPackage = true; }
+      { name = "rockyou"; expressionPath = ./rockyou.nix; }
+      { name = "seclists"; expressionPath = ./seclists.nix; }
+      { name = "wfuzz"; expressionPath = ./wfuzz.nix; needsPackage = true; }
+    ];
+
+  scopedWordlists = lib.makeScope pkgs.newScope (self: with self;
+    let _packageWithLinkCommandFrom = acc: datum:
+      # Some wordlists are based on an already existing `nixpkgs` package; Pass it as parameter of the expression.
+      let parameters = {}
+        // lib.optionalAttrs ((builtins.elem "needsPackage" (builtins.attrNames datum)) && datum.needsPackage) { ${datum.name} = pkgs.${datum.name}; };
+      in
+        let p = callPackage datum.expressionPath parameters;
+        in acc // { ${datum.name} = p; };
+    in builtins.foldl' _packageWithLinkCommandFrom {} _wordlistData
+  );
```
The only thing I don't really like about this PR is this big expression, which can be rather easily simplified.

Pamplemousse replied:
@zseri This kind of review is not helpful at best.

I have tried my best to make this expression as elegant in nix while still respecting the way I wanted to use it.
For that, I had to dig into the documentation, and bang my head at walls to try to follow advice that was given along the way (I did not know everything right of the bat).
It's fine, I was happy to do it, I learned stuff along the way, and I understand there are quality requirements for my work to match the expectations and needs of the community.

> rather easily simplified

Please enlighten us...
Sarcasm aside, the use of "easily" is unnecessary, and can be quite upsetting for those of us who do not find that "easy" (https://css-tricks.com/words-avoid-educational-writing/#10-easy).

So, unless there is something specific (and pertinent) that can be pointed out, and some kind of guidance about how to fix it: :shrug:

zseri replied:
ok, sorry for my bluntness... here's a patch: https://github.com/zseri/nixpkgs/commit/3667eb616bbd51ea7ee1318c09317a38e85c86e2.
Tested with: `nix-build -I nixpkgs=./. ../nixpkgs-wordlists-test.nix`
```nix
# ../nixpkgs-wordlists-test.nix
{ nixpkgs ? <nixpkgs>, system ? builtins.currentSystem }:
(import nixpkgs { inherit system; }).pkgs.security-wordlists.withLists (ps: with ps; [ dirbuster nmap ])
```

Pamplemousse replied:
Thanks for sharing this!
Indeed, this looks much less convoluted.

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,68 @@
+    in builtins.foldl' _packageWithLinkCommandFrom {} _wordlistData
+  );
+
+in let
```
I suppose this line is unnecessary

zseri commented on pkgs/tools/security/wordlists/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+      paths = wordlistPackages;
+
+      dontUnpack = true;
```
it should be documented what this line does.

Pamplemousse replied:
I don't really know which line this comment applies to...
If it is about `dontUnpack`, it is documented at https://nixos.org/manual/nixpkgs/stable/#ssec-unpack-phase; If it is about `paths`, then it is a parameter of the `symlinkJoin` function used, and it is documented at https://nixos.org/manual/nixpkgs/stable/#chap-trivial-builders .

zseri replied:
other `runCommand` invocations (`symlinkJoin` also does this) don't do this, e.g. https://github.com/NixOS/nixpkgs/blob/dfcfe50920c81302be069533514a16d4a48c6715/pkgs/build-support/trivial-builders.nix#L255-L271; therefore I'm not sure if it is really necessary.

zseri replied:
I tested it without it and couldn't detect a difference in the output, therefore I assume that it isn't necessary.

fogti reviewed (approved):
otherwise fine

Pamplemousse committed:
security-wordlists: Run `nixpkgs-fmt`

Signed-off-by: Pamplemousse <xav.maso@gmail.com>

fogti committed:
security-wordlists: Refactoring

zseri commented:
my overall remaining suggestions: https://github.com/zseri/nixpkgs/tree/wordlists / https://github.com/zseri/nixpkgs/commit/9af53070206597697667f3a6550ce03829519bab

fogti reviewed (changes requested):
please try to incorporate my newly suggestions; I rebased them. https://github.com/zseri/nixpkgs/tree/wordlists / https://github.com/Pamplemousse/nixpkgs/pull/1

zseri commented:
currently, `makeScope` is used (probably to make extending the package set easier), but extending the package set of wordlists is still not easy, because it involves `overrideAttrs` on `withLists` to be comfortably usable. Additionally, I'm unsure if we should provide an attribute `withLists'` which takes the list of packages directly instead of a function.

fogti reviewed (approved):
anyway, I would really appreciate if this PR could be merged rather sooner than later, it doesn't matter (to me) if my suggestions get merged into this PR, or if I try to get them in via the usual nixpkgs PR workflow after this has been merged. I think the interface is good enough, it works, details and interna can be improved later on / in the long run.

Pamplemousse commented:
@zseri, thanks, it took your comment on the other PR for me to realise I did not push the commit I took from what you did...

zseri commented:
thanks.

@SuperSandro2000 please re-review. this imo is now ready to be merged.

iblech commented:
Looks good also to me!

buckley310 commented:
Was putting the wordlists in a sub folder `$out/share/wordlists/` considered? I don't feel too strongly either way, but it would be a little more similar to the security distributions (kali) that way.

Pamplemousse commented:
I am tired of this PR, it:
  * is almost a year old
  * has attracted two main rounds of reviews, and corrections were made as soon as possible
  * has been approved several times
  * regularly goes into limbo for several months
  * attracts new kind of nitpicking every now and then

IMHO, this was an improvement on having nothing at all, and good enough to be iterated from if needed.
Closing given that this is unlikely to be merged ever.

jonringer commented:
@Pamplemousse sorry, I think it would be beneficial to have them available in nixpkgs. But I agree, it can be painful to upstream these, especially if you're treading new water.

I would recommend creating a personal repo (or NUR repo) where this can be available.

For me personally, I don't really know how to "ensure this is the correct way forward"... not to mention it would take me to ~14 hours a day to tackle all of my github notifications a day :(
